cssselect/xpath.py

```diff
@@ -643,25 +678,31 @@ def pseudo_never_matches(self, xpath):
 
     # Attrib: dispatch by attribute operator
 
-    def xpath_attrib_exists(self, xpath, name, value):
+    def xpath_attrib_exists(self, xpath: XPathExpr, name: str, value: Optional[str]) -> XPathExpr:
         assert not value
         xpath.add_condition(name)
         return xpath
 
-    def xpath_attrib_equals(self, xpath, name, value):
+    def xpath_attrib_equals(self, xpath: XPathExpr, name: str, value: Optional[str]) -> XPathExpr:
+        assert value
         xpath.add_condition("%s = %s" % (name, self.xpath_literal(value)))
         return xpath
 
-    def xpath_attrib_different(self, xpath, name, value):
+    def xpath_attrib_different(
+        self, xpath: XPathExpr, name: str, value: Optional[str]
+    ) -> XPathExpr:
+        assert value
```

Comment:
This assert looks wrong as the code below has a `if value` condition

Reply:
Thoug it's not really clear to me how does that code work. I assume `value` can be an empty string but cannot be None? In any case it makes no sense to call xpath_literal() on False or None, and for an empty string the result could be hardcoded instead.

Reply:
Fun fact: `[attribute!=value]` is a cssselect-specific extension which is not in any CSS standard. This also means its behavior is not described anywhere. So let's look how it works:

```
In [3]: GenericTranslator().css_to_xpath('[foo!=bar]')
Out[3]: "descendant-or-self::*[not(@foo) or @foo != 'bar']"

In [5]: GenericTranslator().css_to_xpath('[foo!=""]')
Out[5]: "descendant-or-self::*[@foo != '']"
```

So valid matches for non-empty values are an absent attribute and an attribute that isn't equal to the value. The only valid match for an empty value is a present non-empty attribute. We should keep this behavior.

This also applies to `xpath_attrib_equals` above, where value can't be None but can be an empty string.